apps/btc_family/btc_txn.c

```diff
@@ -247,31 +286,183 @@ static bool fetch_transaction_meta(btc_query_t *query) {
          sizeof(btc_sign_txn_metadata_t));
-  btc_txn_context->inputs = (sign_txn_input_t *)malloc(
-      sizeof(sign_txn_input_t) * btc_txn_context->metadata.input_count);
+  btc_txn_context->inputs = (btc_txn_input_t *)malloc(
+      sizeof(btc_txn_input_t) * btc_txn_context->metadata.input_count);
   btc_txn_context->outputs = (btc_sign_txn_output_t *)malloc(
       sizeof(btc_sign_txn_output_t) * btc_txn_context->metadata.output_count);
   // TODO: check if malloc failed; report to host and exit
   send_response(BTC_SIGN_TXN_RESPONSE_META_ACCEPTED_TAG);
   return true;
 }
 
+static bool fetch_input_utxo(btc_query_t *query) {
+  // Validate input UTXOs for safety from attack. Ref:
+  // https://blog.trezor.io/details-of-firmware-updates-for-trezor-one-version-1-9-1-and-trezor-model-t-version-2-3-1-1eba8f60f2dd
+  for (int idx = 0; idx < btc_txn_context->metadata.input_count; idx++) {
+    if (!btc_get_query(query, BTC_QUERY_SIGN_TXN_TAG) &&
+        !check_which_request(query, BTC_SIGN_TXN_REQUEST_INPUT_TAG)) {
+      return false;
+    }
+    // TODO: check input UTXO type; exit if unsupported
+    // P2PK 68, P2PKH 25 (21 excluding OP_CODES), P2WPKH 22, P2MS ~, P2SH 23 (21
+    // excluding OP_CODES) refer https://learnmeabitcoin.com/technical/script
+    // for explaination Currently the device can spend P2PKH or P2WPKH UTXOs
+    // only for (int i = 0; i < in_count; i++) {
+    //   if (txn_ctx->inputs[i].script_pub_key.size != 22 &&
+    //       txn_ctx->inputs[i].script_pub_key.size != 25) {
+    //     return false;
+    //   }
+    // }
+
+    // verify transaction details and discard the raw-transaction (prev_txn)
+    const btc_sign_txn_input_prev_txn_t *txn = &query->sign_txn.input.prev_txn;
+    if (!btc_verify_input_utxo(
+            txn->bytes, txn->size, &btc_txn_context->inputs[idx])) {
+      // input UTXO validation failed, terminate immediately
+      btc_send_error(ERROR_COMMON_ERROR_CORRUPT_DATA_TAG,
+                     ERROR_DATA_FLOW_INVALID_DATA);
+      return false;
+    }
+
+    // clone the input UTXO details into btc_txn_context
+    btc_txn_input_t *input = &btc_txn_context->inputs[idx];
+    input->prev_output_index = query->sign_txn.input.prev_output_index;
+    input->address_index = query->sign_txn.input.address_index;
+    input->change_index = query->sign_txn.input.change_index;
+    input->value = query->sign_txn.input.value;
+    input->sequence = query->sign_txn.input.sequence;
+    input->script_pub_key.size = query->sign_txn.input.script_pub_key.size;
+    memcpy(input->prev_txn_hash, query->sign_txn.input.prev_txn_hash, 32);
+    memcpy(input->script_pub_key.bytes,
+           query->sign_txn.input.script_pub_key.bytes,
+           input->script_pub_key.size);
+
+    // send accepted response to indicate validation of input passed
+    send_response(BTC_SIGN_TXN_RESPONSE_INPUT_ACCEPTED_TAG);
+  }
+  return true;
+}
+
+static bool fetch_output_utxo(btc_query_t *query) {
```

Comment:
While an output of an unsigned transaction will become a utxo once it gets added to the blockchain. It should not be called as utxo here as this leads to some confusion.

To avoid this please refer to [official bitcoin reference glossary](https://developer.bitcoin.org/glossary.html)
Relevant searches: TxOut, TxIn, Change Output, UTXO

Reply:
now using output and input instead and removed the UTXO term